variety/__init__.py

```diff
@@ -87,18 +90,169 @@ def _get_dbus_key():
-        return "" if result is None else result
-
+class IVarietyService():
+    def __init__(self):
+        pass
+
+    # Command to start the listener thread (or do nothing if not singleton)
+    def start_listener(self):
+        pass
+
+    # Send command to the listener thread of the master process
+    def send_command(self):
+        pass
+
+    # Set the variety window for this class
+    def set_variety_window(self, variety_window):
+        pass
+
+    # Return whether or not this service is the current master service
+    def is_master(self):
+        pass
+
+if os.name == 'nt':
+    import asyncio
+    import threading
+    import pickle
+    class VarietyService(IVarietyService, threading.Thread):
+        def __init__(self):
+            super(IVarietyService, self).__init__()
+            super(threading.Thread, self).__init__()
+            self.loop = asyncio.get_event_loop()
+
+            # Fire up command server thread
+            self.server_thread = self.CommandServerThread()
+            self.server_thread.set_variety_service(self)
+            self.server_thread.start()
+
+            # Will exit as soon as thread exits if could not bind
+            self.server_thread.join(timeout=2)
+            # If we can't bind, this is a client and thread will be dead.
+            self.master = self.server_thread.is_alive()
+
+        @staticmethod
+        def unix_sockets_supported():
+            # Maybe some day for windows: https://bugs.python.org/issue33408
+            return hasattr(socket, 'AF_UNIX')
+
+        @staticmethod
+        def get_sock_file():
+            return os.path.join(os.path.expanduser("~/.config/variety/"), "ipc.sock")
+
+        @staticmethod
+        def start_server(cb, loop):
+            if not VarietyService.unix_sockets_supported():
+                return asyncio.start_server(cb, '127.0.0.1', 8888, loop=loop)
+            else:
+                return asyncio.start_unix_server(cb, '127.0.0.1', path=VarietyService.get_sock_file(), loop=loop)
+
+        @staticmethod
+        def open_connection(loop):
+            if not VarietyService.unix_sockets_supported():
+                return asyncio.open_connection('127.0.0.1', 8888, loop=loop)
+            else:
+                return asyncio.open_unix_connection(path=VarietyService.get_sock_file(), loop=loop)
+
+        class CommandServerThread(threading.Thread):
+            def set_variety_service(self, service):
+                self.service = service
+
+            def run(self):
+                self.coro = VarietyService.start_server(self.process_command_server, loop=self.service.loop)
+                try:
+                    self.server = self.service.loop.run_until_complete(self.coro)
+                    # Serve requests until Ctrl+C is pressed
+                    try:
+                        self.service.loop.run_until_complete(self.server.wait_closed())
```

Comment:
FWIW I'm not sure how safe it is to mix threads and asyncio code. It might beeasier to just stick to one concurrency concept at a time?